File: php/data_structures/linked_list.php
<?php
class LinkNode{
    /**
     * @var mixed
     */
    public $value;
    /**
     * @var LinkNode | NULL
     */
    public $next;
    public function __construct($val, $nxt=NULL){
        $this->value = $val;
        $this->next = $nxt;
    }
}
class LinkedList{
    /**
     * @var LinkNode | NULL
     */
    private ?LinkNode $head_;
    /**
     * @var LinkNode | NULL
     */
    private ?LinkNode $tail_;
    /**
     * @var int
     */
    private int $length_;

    public function __construct() {
        $this->head_ = NULL;
        $this->tail_ = NULL;
        $this->length_ = 0;
    }

    public function get($index): ?LinkNode {
        if ($index < 0 or $index > $this->length_){
            return NULL;
        }
        $cur = $this->head_;
        $i = 0;
        while ($i != $index){
            $cur = $cur->next;
            $i++;
        }
        return $cur;
    }
	public function unshift($value){
    	if($this->head_)
    		$this->head_ = new LinkNode($value,$this->head_);
    	else{
    		$this->head_ = new LinkNode($value);
    		$this->tail_ = $this->head_;
		}
    	$this->length_++;
	}
	public function shift(): ?LinkNode {
    	if($this->head_){
    		$tmp = $this->head_;
    		$this->head_ = $tmp->next;
    		$this->length_--;
    		if($this->length_ === 0){
    			$this->tail_ = NULL;
			}
    		return $tmp;
		}
    	return NULL;
	}

    public function insert($index,$value){
    	if($index > $this->length_ || $index < 0)
    		return NULL;
    	else if($index == 0)
    		$this->unshift($value);
    	else if($this->head_ === null) {
			$this->head_ = new LinkNode($value);
			$this->tail_ = $this->head_;
		}
    	else{
    		$prev = $this->get($index - 1);
    		$prev->next = new LinkNode($value,$prev->next);
		}
    	$this->length_++;
	}
	public function set($index, $value){
    	if($index > 0 && $index < $this->length_){
    		$p = $this->get($index);
    		$p->value = $value;
		}
	}
    /**
     * @return LinkNode|null
     */
    public function pop(): ?LinkNode {
        if($this->head_ == NULL){
            return NULL;
        }
        $popped = $this->tail_;
        if($this->length_ > 1) {
			$prev = $this->get($this->length_-2);
			$this->tail_ = $prev;
			$this->tail_->next = NULL;
		}
        else
	        $this->tail_ = NULL;
        $this->length_--;
        return $popped;
    }
	public function remove($index): ?LinkNode {
		if($index < 0 || $index > $this->length_)
			return NULL;

		if($index === $this->length_ - 1)
			return $this->pop();
		else if($index === 0)
			return $this->shift();
		else{
			$prev = $this->get($index-1);
			$removed = $prev->next;
			$prev->next = $removed->next;
			$this->length_--;
			return $removed;
		}
	}
    public function push($value){
        if($this->head_ == NULL){
            $this->head_ = new LinkNode($value);
            $this->tail_ = $this->head_;
            $this->length_++;
        }
        else{
            $cur = $this->head_;
            while($cur->next != NULL){
                $cur = $cur->next;
            }
            $cur->next = new LinkNode($value);
            $this->tail_ = $cur->next;
            $this->length_++;
        }
    }

    public function reverse(){
    	$current = $this->head_;
    	$prev = NULL;
    	$next = NULL;
    	$this->head_ = $this->tail_;
    	$this->tail_ = $current;
    	while($current){
    		$next = $current->next;
    		$current->next = $prev;
    		$prev = $current;
    		$current = $next;
		}
		$this->head_ = $prev;
	}
    public function __toString(): string {
        $cur = $this->head_;
        $res = "[";
        while($cur->next != NULL){
            $res.="$cur->value, ";
            $cur = $cur->next;
        }
        $res .= "$cur->value]";
        return $res;
    }

}

if (!count(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS))){
	$ll = new LinkedList();
	$ll->push(4);
	$ll->push(5);
	$ll->push(6);
	$ll->insert(2, 222);
	$ll->push(7);
	$ll->push(8);
	$ll->push(9);
	$ll->push(10);
	$ll->reverse();
	$ll->pop();
	$ll->set(2, 100);
	$ll->insert(5, 555);
	$ll->insert(4, 444);
	$ll->shift();
	$ll->unshift(6);
	$ll->remove(5);
	$expected = '[6, 9, 100, 7, 444, 555, 222, 5]';
	if((string)$ll === $expected)
		echo "All's Clear\n";
	else
		echo "Error\n$ll\n";
}

File: lua/data_structures/linked_list.lua
--uncomment if you haven't set your working directory to the starting folder
--package.path = package.path .. ";../?.lua"
require 'util'
--[[
Methods


Note: The indexes are based on 1 as this is lua
]]

--[[
]]

-- Linked List Class
LinkedList = Class{}

function LinkedList:init()
	self.Node = Class{}
	function self.Node:init(value,ref)
		self.value = value
		self.next = ref
	end
	self._head = nil
	self._length = 0
	self._tail = nil
end

-- Get Method
-- @param idx The index of the item to get
function LinkedList:get(idx)
	if idx < 1 or idx > self._length then
		return nil
	end

	local cur = self._head

	for i=1,idx-1 do
		cur = cur.next
	end
	return cur
end


-- Adds a value to the end of the list
function LinkedList:push(value)
	if self._head == nil then
		self._head = self.Node(value)
		self._tail = self._head
	else
		local cur = self._head
		while(cur.next ~= nil) do
			cur = cur.next
		end
		cur.next = self.Node(value)
		self._tail = cur.next
	end
	self._length = self._length +  1
end

-- Removes an item from the end of the list or the one given by the index
function LinkedList:pop(idx)
	if self._head == nil then
		return nil
	end
	local popped
	if idx then
		local prev = self:get(idx-1)
		popped = prev.next
		prev.next = prev.next.next
	else
		popped = self._tail.value
		if self._length > 1 then
			local prev = self:get(self._length-1)
			self._tail = prev
			self._tail.next = nil
		else
			self._tail = nil
		end
	end
	self._length = self._length - 1
	return popped
end


-- reverse the list in place
function LinkedList:reverse()
	local cur = self._head
	self._head = self._tail
	self._tail = cur
	local prev = nil

	while(cur ~= nil) do
		local next = cur.next
		cur.next = prev
		prev = cur
		cur = next
	end
	self._head = prev
end


-- insert a value at specified index
function LinkedList:insert(idx,value)
	if 1 > idx or idx > self._length then
		return nil
	end
	local prev = self:get(idx)
	prev.next = self.Node(value, prev.next)
	self._length = self._length + 1
end


-- change the value of the node at the given index to the value given
function LinkedList:set(idx,value)
	if idx < 1 or idx > self._length then
		return nil
	end
	local Node = self:get(idx)
	Node.value = value
end


-- get the length of the list
function LinkedList:len()
	return self._length
end

-- remove the head node
function LinkedList:shift()
	if self._head then
		local tmp = self._head
		self._head = tmp.next
		self._length = self._length - 1
		if self._length == 0 then
			self._tail = nil
		end
		return tmp
	else
		return nil
	end
end

-- remove an item at the given index
-- @param idx The index
function LinkedList:remove(idx)
	if 1 < idx or idx > self._length then
		return nil
	end
	if idx == self._length - 1 then
		self:pop()
	elseif idx == 0 then
		self:shift()
	else
		local prev = self:get(idx-1)
		local removed = prev.next
		prev.next = removed.next
		self._length = self._length - 1
	end
end

-- add an element to list before the current head
function LinkedList:unshift(value)
	if self._head then
		self._head = self.Node(value, self._head)
	else
		self._head = self.Node(value)
		self._tail = self._head
	end
end

-- convert it into a normal "list"
function LinkedList:stringify()
	local os = '['
	local cur = self._head
	while cur.next ~= nil do
		os = os .. tostring(cur.value) .. ","
		cur = cur.next
	end
	return os .. tostring(cur.value) .. "]"
end
if not pcall(debug.getlocal,4,1) then
	ll = LinkedList()
	ll:push(1)
	ll:push(2)
	ll:push(3)
	ll:push(4)
	ll:push(5)
	ll:push(6)
	print(ll:stringify())
	ll:pop()
	ll:pop()
	print(ll:stringify())
	ll:shift()
	ll:unshift(11)
	print(ll:stringify())
	ll:pop(2)
	print(ll:stringify())
	-- nil test
	print(ll:get(0))
	ll:reverse()
	print(ll:stringify())
end